### ISSUE TITLE
Bolt connection in 3.1.0-M08 is incorrectly closed after INIT failure

### ISSUE THREAD

**Issue by cleishm** [neo4j/neo4j#7955]
- Neo4j version: 3.1.0-M08
- API/Driver: Bolt

In 3.0.4:

```
DEBUG [connection]: attempting connection to ::1 [7687]
TRACE [connection]: opened socket to localhost [7687] (fd=6)
TRACE [connection]: initialiting TLS (fd=6)
DEBUG [tls]: server cert fingerprint: xxx
DEBUG [tls]: TLS certificate verification failed: self signed certificate
INFO  [connection]: connected (0x7fc969419590) to localhost:7687
DEBUG [connection]: connection 0x7fc969419590 using protocol version 1
DEBUG [session]: new session (0x7fc969504230) on 0x7fc969419590
TRACE [session]: enqu INIT{"libneo4j-client/1.2.2~devel", {scheme: basic, principal: "neo4j", credentials: ****}} (0x7fc969806a00) in 0x7fc969504230
DEBUG [session]: sent INIT (0x7fc969806a00) in 0x7fc969504230
DEBUG [session]: rcvd FAILURE in response to INIT (0x7fc969806a00)
TRACE [session]: FAILURE in 0x7fc969504230 (response to INIT): {code:"Neo.ClientError.Security.Unauthorized",message:"The client is unauthorized due to authentication failure."}
TRACE [session]: enqu ACK_FAILURE (0x7fc969806ae0) in 0x7fc969504230
DEBUG [session]: sent ACK_FAILURE (0x7fc969806ae0) in 0x7fc969504230
DEBUG [session]: rcvd SUCCESS in response to ACK_FAILURE (0x7fc969806ae0)
TRACE [session]: ACK_FAILURE complete in 0x7fc969504230
```

In 3.1.0-M08

```
DEBUG [connection]: attempting connection to ::1 [7687]
INFO  [connection]: connection to ::1 [7687] failed: Connection refused
DEBUG [connection]: attempting connection to 127.0.0.1 [7687]
TRACE [connection]: opened socket to localhost [7687] (fd=6)
TRACE [connection]: initialiting TLS (fd=6)
DEBUG [tls]: server cert fingerprint: xxx
DEBUG [tls]: TLS certificate verification failed: self signed certificate
INFO  [connection]: connected (0x7fbae3d11e50) to localhost:7687
DEBUG [connection]: connection 0x7fbae3d11e50 using protocol version 1
DEBUG [session]: new session (0x7fbae3d11650) on 0x7fbae3d11e50
TRACE [session]: enqu INIT{"libneo4j-client/1.2.2~devel", {scheme: basic, principal: "neo4j", credentials: ****}} (0x7fbae4809e00) in 0x7fbae3d11650
DEBUG [session]: sent INIT (0x7fbae4809e00) in 0x7fbae3d11650
DEBUG [session]: rcvd FAILURE in response to INIT (0x7fbae4809e00)
TRACE [session]: FAILURE in 0x7fbae3d11650 (response to INIT): {code:"Neo.ClientError.Security.Unauthorized",message:"The client is unauthorized due to authentication failure."}
TRACE [session]: enqu ACK_FAILURE (0x7fbae4809ee0) in 0x7fbae3d11650
DEBUG [session]: sent ACK_FAILURE (0x7fbae4809ee0) in 0x7fbae3d11650
TRACE [session]: neo4j_connection_recv failed: Connection closed
TRACE [session]: draining ACK_FAILURE (0x7fbae4809ee0) from queue on 0x7fbae3d11650
connection failed: Connection closed
INFO  [connection]: disconnected 0x7fbae3d11e50
```

Note the failure at `TRACE [session]: neo4j_connection_recv failed: Connection closed`, indicating the connection was closed unexpectedly before an `ACK_FAILURE` could be delivered.

This breaks the flow of the protocol and causes a regression with existing tooling like `neo4j-client`.

**Comment by cleishm** [comment]
Note also that existing tools depend on the protocol state returning to pre-`INIT` after the ACK_FAILURE is sent, allowing for `INIT` to be re-attempted.

**Comment by technige** [comment]
This undocumented behaviour was removed as part of a broader hardening and simplification of the server implementation. You should now consider closing after INIT failure to be correct behaviour. To retry, simply create a new connection.

The reasoning behind this change is threefold:
1. To bring the server implementation more closely in line with the facilities provided by the drivers, which expose no specific retry API.
2. To reduce the number of potential attack vectors for malicious client software by reducing the number of pre-auth states within the state machine.
3. Removing this is no way limits functionality. Any software that wants to try a different password simply has to recreate the `Driver` object or, if working at a lower level, open a new connection.

**Comment by cleishm** [comment]
This was discussed in a previous ticket (#7356), which was resolved with the behaviour as stated. This is effectively as good as documented. Additionally, there is now released software depending on this behaviour, which cannot be reverted or replaced easily.

If you intend to make such changes, I suggest bumping the protocol version of Bolt and supporting the previous version (for at least one major release).

**Comment by technige** [comment]
Thank you for your feedback.

This behaviour is not referenced from any official Bolt documentation. GitHub discussions should not be considered as such. Therefore, enforcing this limitation within the 3.1 Bolt server implementation does not require a version bump for the protocol as the protocol itself has not evolved.

I'm unaware of anyone else who relies on this implementation-specific behaviour, since most users either use the official drivers or use a community driver based on the official drivers, which don't offer auth retries. Ultimately, this benefits more users than it disadvantages.

Note that you do have a full workaround available. Simply handle the disconnection and establish a new connection on login failure. This will work perfectly for both 3.0 and 3.1 servers.

**Comment by cleishm** [comment]
Unfortunately, the proposed "workaround" requires modifying already released software. This software has been packaged into distributions such as Ubuntu, and Ubuntu has been frozen before the upcoming stable release. Thus changing the code is no longer possible.

Note that the behaviour was also documented in integration tests, https://github.com/neo4j/neo4j/blob/3.0/community/bolt/src/test/java/org/neo4j/bolt/v1/runtime/integration/SessionAuthIT.java#L162-L212. Considering that the only documentation for the BOLT protocol exists within the neo4j codebase itself, that would appear to be a something reasonable to rely on.

It's great that you want to make changes to the protocol, but those should only be made in a new protocol version, with support for the previous continuing until at least the next major release (neo4j 4.0). It's also a substantive change, as it collapses the separate concept of a session from the underlying connection and the driver model should be updated to reflect that.

**Comment by technige** [comment]
'The proposed "workaround"' is relevant only if you also want to support version 3.1 of the server after its release. You should be able to rely on the retry being available in 3.0 as you can today.

Any already-released software cannot realistically not be built to support other software that has not yet been released. Changing the protocol version would not alter that reality.

There is no plan to reintroduce this for 3.1.

**Comment by cleishm** [comment]
I'm sorry, but that is simply not an acceptable response.

A client released in a stable distribution of a well used operating system (Ubuntu) needs to rely on the same protocol version not changing, regardless of new server releases. There is no opportunity to fundamentally change the client in a stable release.

If you want to change the protocol, then there is a protocol version negotiation that you've added for exactly that reason, and it should go through a process of deprecation. Especially as there is no way to introspect what version of Neo4j the server is running before authentication, and regardless there is no way to update the client software to do so anyway (it's already released).
